File: src/main/kotlin/com/lwyr/ai/service/AuthService.kt
package com.lwyr.ai.service

import com.lwyr.ai.dto.auth.LoginRequest
import com.lwyr.ai.dto.auth.SignupRequest
import com.lwyr.ai.entity.User
import com.lwyr.ai.entity.UserRole
import com.lwyr.ai.exception.AuthenticationException
import com.lwyr.ai.exception.ResourceAlreadyExistsException
import com.lwyr.ai.repository.UserRepository
import mu.KotlinLogging
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import java.time.OffsetDateTime
import java.util.UUID

private val logger = KotlinLogging.logger {}

@Service
class AuthService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder
) {

    fun signup(request: SignupRequest): User {
        logger.info { "Processing signup for email: ${request.email}" }

        if (userRepository.existsByEmail(request.email)) {
            logger.warn { "Signup failed - email already exists: ${request.email}" }
            throw ResourceAlreadyExistsException("Email already registered")
        }

        val now = OffsetDateTime.now()
        val user = User(
            id = UUID.randomUUID(),
            email = request.email,
            passwordHash = passwordEncoder.encode(request.password),
            role = UserRole.USER,
            createdAt = now,
            updatedAt = now
        )

        val savedUser = userRepository.save(user)
        logger.info { "User registered successfully: ${savedUser.id}" }

        return savedUser
    }

    fun login(request: LoginRequest): User {
        logger.info { "Processing login for email: ${request.email}" }

        val user = userRepository.findByEmail(request.email)
            ?: run {
                logger.warn { "Login failed - user not found: ${request.email}" }
                throw AuthenticationException("Invalid email or password")
            }

        if (!passwordEncoder.matches(request.password, user.passwordHash)) {
            logger.warn { "Login failed - invalid password for email: ${request.email}" }
            throw AuthenticationException("Invalid email or password")
        }

        logger.info { "Login successful for user: ${user.id}" }
        return user
    }
}

File: app/services/auth_service.py
"""Authentication service."""

import logging
from uuid import UUID

from sqlalchemy.orm import Session

from app.exceptions import AuthenticationException, ResourceAlreadyExistsException
from app.models import UserRole
from app.repositories import UserRepository
from app.security import create_access_token, hash_password, verify_password

logger = logging.getLogger(__name__)


class AuthService:
    """Service for authentication operations."""

    def __init__(self, db: Session) -> None:
        self.db = db
        self.user_repository = UserRepository(db)

    def signup(self, email: str, password: str) -> dict:
        """Register a new user."""
        logger.info(f"Processing signup for email: {email}")

        if self.user_repository.exists_by_email(email):
            logger.warning(f"Signup failed - email already exists: {email}")
            raise ResourceAlreadyExistsException("Email already registered")

        password_hash = hash_password(password)
        user = self.user_repository.create(email, password_hash, UserRole.USER)

        logger.info(f"User registered successfully: {user['id']}")

        return user

    def login(self, email: str, password: str) -> dict:
        """Authenticate a user and return user data."""
        logger.info(f"Processing login for email: {email}")

        user = self.user_repository.find_by_email(email)
        if user is None:
            logger.warning(f"Login failed - user not found: {email}")
            raise AuthenticationException("Invalid email or password")

        if not verify_password(password, user["password_hash"]):
            logger.warning(f"Login failed - invalid password for email: {email}")
            raise AuthenticationException("Invalid email or password")

        logger.info(f"Login successful for user: {user['id']}")

        return user

    def generate_token(self, user: dict) -> str:
        """Generate a JWT token for a user."""
        return create_access_token(
            user_id=str(user["id"]),
            email=user["email"],
            role=user["role"].value if isinstance(user["role"], UserRole) else user["role"],
        )

    def get_user_by_id(self, user_id: UUID) -> dict | None:
        """Get a user by ID."""
        return self.user_repository.find_by_id(user_id)
